### ISSUE TITLE
[feature request]: convertFromHTML support custom entity.

### ISSUE THREAD

**Issue by RaoHai** [facebook/draft-js#685]
https://github.com/facebook/draft-js/blob/master/src/model/encoding/convertFromHTMLToContentBlocks.js#L408

During this loop, draft-js should create a `LINK` typed entity if this child has `href` attribute and has valid link text.

We would like to handle the condition and create a custom entity (e.g., convert `<img />` tag to an `IMAGE` entity ), just like :

```javascript
while (child) {
  if (this.props.ishandleHTML
    && entity = this.props.handleParseHTMLElement(child)
  ) {
    newEntityMap = addEntityToEntityMap(
        newEntityMap,
        entity
    );
    entityId = newEntityMap.keySeq().last();
 } else {
 // original logic...
}
```

**Comment by N1kto** [comment]
I am afraid you will first face a problem of 'capturing' `<img />` elements from the pasted HTML string. See #523 .

If that is not a problem, i.e. you get your image blocks in the output array from `convertFromHTML`, you can easily update `entityData` by finding/filtering such blocks from the output array and altering the data using `Entity` API or replacing such block with a new `ContentBlock` instance.

**Comment by RaoHai** [comment]
@N1kto yes, I've faced the problem of parsing `<img />` tag, but not totally the same as yours ..

I think your problem can be solved by passing `blockRenderMap` props, 
e.g. 
```javascript
convertFromHTML('<video src="movie.ogg" controls="controls"> Browser not support</video>', 
        getSafeBodyFromHTML, 
        DefaultDraftBlockRenderMap.merge(Map({
          'video': {
            element: 'video'
          },
      )))
```

But my problem is:  
```javascript
convertFromHTML('<img src="x.jpg" />', 
        getSafeBodyFromHTML, 
        DefaultDraftBlockRenderMap.merge(Map({
          'image': {
            element: 'img'
          },
      )))
```
cannot get video typed block,  because child of an `<img />` tag is null, I think.

`https://github.com/facebook/draft-js/blob/master/src/model/encoding/convertFromHTMLToContentBlocks.js#L407`

**Comment by N1kto** [comment]
You are right regarding my problem.
Regarding the issue subject - altering entity once you have a `ContentBlock` is not a big deal, imho, so such feature is not that demanded.

The real blocking point in your case is that you cannot parse `<img />` into `ContentBlock`.

If you have control over the source HTML you can try replacing `<img />` occurrences with some other HTML element with children and then alter it in contentBlocks array. Sounds kind a hack, but this is how I was converting images we had in our old editor as simple HTML markup into DraftJS `atomic` block with their further rendering as react component.

**Comment by michelson** [comment]
@N1kto can you provide and example of it ? what kind of tag are you using when replace `<img>` ?

also i'm trying to run this , in draft 0.9.1:

```
convertFromHTML('<video src="movie.ogg" controls="controls"> Browser not support</video>', 
        getSafeBodyFromHTML, 
        DefaultDraftBlockRenderMap.merge(Map({
          'video': {
            element: 'video'
          }
      })))
```
but I get `unstyled` type blockMap not video type as @RaoHai said

any help will be greatly appreciated

best!

**Comment by N1kto** [comment]
Hard to say why Draft doesn't handle `<video />` blocks in your case. It's rather 'special' element whose children are not text nodes, maybe that causes Draft to fall back to `unstyled` blocks.
Never the less a dirty hack I used might be helpful in your case. It's pretty dump, but it works.
Please refer to this [gist](https://gist.github.com/N1kto/6702e1c2d89a33a15a032c234fc4c34e). I 'transpiled' it from our project code which is written in `coffeescript`, so there can be typos/minor mistakes, just debug it in that case. It works for us - we convert images and custom tables (tables generated based on internal data with custom attributes) and it works fine.

The idea is to collect all required attributes from target element, e.g. `src` attribute of `<img />` element (which would enable us to render it later with React component) to JSON, then stringify it and save as simple `textContent` within some "temporary" text block element which is recognizable by Draft, in mentioned gist it is `<blockquote />`. It has to be some element which you don't normally use in your editor implementation to avoid collisions. I took `<blockquote />` because we don't use such element in our editor and it is recognizable by Draft out of the box, so no need to bother with custom `blockRenderMap`.
Then after `convertFromHTML` successfully converted your content you replace `<blockquote />` blocks with `atomic` ones, you parse JSON string into object and inject that data in Entity (or in block meta data - it's implementation details). Later you render such `atomic` blocks with React component (`<img />` src you will get from Entity data.
Hope that helped.

**Comment by michelson** [comment]
Hi @N1kto , the dirty hack works great! very wise move 👍 
thanks again for your detailed explanation 
best

**Comment by gbbr** [comment]
@N1kto very nice! You just saved me a lot of trouble. Thanks!!!

I wonder how we can create these entities for the character list once 0.11 is out and the Entity API gets wiped.

**Comment by N1kto** [comment]
@gbbr thanks for the heads up. Didn't follow DraftJS for a while, good to know and get ready for API changes.
Need to study API changes to give a good answer, but the quick answer to your question is (also used this approach in our code) following: don't use `Entity`, but use `ContentBlock`'s `data` prop in combination with custom block type.
So, for example, this part from mentioned gist:
```
    const { imgSrc } = JSON.parse(block.getText())
    const entityData = {
      type: 'IMAGE',
      mutability: 'IMMUTABLE',
      data: { imgSrc }
    }
    const blockSpec = extend({ type: 'atomic', text: ' ' }, { entityData })
```
Should be replaced with this:
```
    const { imgSrc } = JSON.parse(block.getText())
    const blockSpec = { type: 'custom_type_name_it_whatever', data: { imgSrc } }
```
Of course, in `createContentBlock` you will also no longer need `Entity`-related code.
So, as you see we store parsed data in `data` prop of the `ContentBlock`, which is part of `ContentBlock` API, also we set `type ` to _custom_type_name_it_whatever_. And since we added custom type, we need to do 2 more things to get all working.
1. Need to update/set `blockRenderMap` `Editor`'s prop with our custom type
2. Need to update/create `blockRendererFn` `Editor`'s prop which will know how to render our custom (in this case it's `<image />`) block with help of `React`.

It can be something like this:
```
const updatedRenderMap = Draft.DefaultDraftBlockRenderMap.merge(Immutable.Map({
  'custom_type_name_it_whatever':  { element: 'div' }
}))
const blockRendererFn = block => {
  if (block.getType() === 'custom_type_name_it_whatever') {
    return {
      component: ReactComponentToRenderImage,
      props: {...}  // arbitrary props to your ReactComponentToRenderImage
   }
  }
}
```
And then you pass both custom `renderMap` and `blockRendererFn` to DraftJS Editor:

`<Editor blockRenderMap={ updatedRenderMap } blockRendererFn={ blockRendererFn } />`

This is very primitive example, but the general idea should be clear and should work with 0.11 as I didn't mentioned any notification about changing `ContentBlock` API and editor component API. Otherwise that will have to be revised :)
